### PR TITLE
[SPARK-48739][SQL] Disable writing collated data to file formats that don't support them in non managed tables

### PR DESCRIPTION
**Pull request by stefankandic**
### What changes were proposed in this pull request?

Disable writing collated types to data sources that don't support them. However, spark managed tables should still work as the schema is in HMS and not in the file itself.

### Why are the changes needed?

Right now, when users write a collated type directly to json, text, orc.. they will not see that collation when reading back.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Added new UTs

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by stefankandic** [comment]
@cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala [hunk 1]
```diff
@@ -536,10 +536,12 @@ case class DataSource(
               dataSource.toString, field)
           }
         }
+        DataSourceUtils.verifyCollations(dataSource, data.schema)
```
Comment (allisonwang-db) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala:539]: Shall we combine this with the `data.schema.foreach` check mentioned above?

**Comment by cloud-fan** [comment]
For internal file source API, I think we can simply update `FileFormat#supportDataType` in certain formats such as CSV to return false for string with collation. So no new API is needed.

For data source v1, we can add a new API to `CreatableRelationProvider`, like `supportsStringCollation`. Ideally this is not needed as we already have `CreatableRelationProvider#supportsDataType`. But string collation is special as it's still `StringType` and existing v1 data sources may mistakenly support it if they do `case _: StringType => true`

**Comment by stefankandic** [comment]
@cloud-fan 
> For internal file source API, I think we can simply update FileFormat#supportDataType in certain formats such as CSV to return false for string with collation. So no new API is needed.

This would mean that we wouldn't be able to create spark managed tables with collations for those formats. Is that something that we want to do?

**Comment by cloud-fan** [comment]
> This would mean that we wouldn't be able to create spark managed tables with collations for those formats. Is that something that we want to do?

To confirm the goal of this PR: we want to have a new API for file sources to indicate that a type is supported only with a catalog? I think we should be more specific about this, as there are many APIs to use a file source:
1. read a path with a user-specified schema
2. write to a path
3. create external table with a path
4. create managed table

**Pull request closed without merging**